src/psyclone/psyir/transformations/intrinsics/mms_base_trans.py

```diff
@@ -0,0 +1,326 @@
+        (array reference, dimension and mask).
+
+        :param node: a minval, maxval or sum intrinsic.
+        :type node: :py:class:`psyclone.psyir.nodes.IntrinsicCall`
+
+        returns: a tuple containing the 3 arguments.
+        rtype: Tuple[py:class:`psyclone.psyir.nodes.reference.Reference`, \
+                     py:class:`psyclone.psyir.nodes.Literal` | \
+                     :py:class:`psyclone.psyir.nodes.Reference`, \
+                     Optional[:py:class:`psyclone.psyir.nodes.node`]]
+
+        '''
+        # Determine the arguments to the intrinsic
+        args = [None, None, None]
+        arg_names_map = {"array": 0, "dim": 1, "mask": 2}
+        for idx, child in enumerate(node.children):
+            if not node.argument_names[idx]:
+                # positional arg
+                args[idx] = child
+            else:
+                # named arg
+                name = node.argument_names[idx].lower()
+                args[arg_names_map[name]] = child
+        array_ref = args[0]
+        dimension_ref = args[1]
+        mask_ref = args[2]
+        return (array_ref, dimension_ref, mask_ref)
+
+    def __str__(self):
+        return (f"Convert the PSyIR {self._INTRINSIC_NAME} intrinsic "
+                "to equivalent PSyIR code.")
+
+    def validate(self, node, options=None):
+        '''Check that the input node is valid before applying the
+        transformation.
+
+        :param node: a Sum, Minval or Maxval intrinsic.
+        :type node: :py:class:`psyclone.psyir.nodes.IntrinsicCall`
+        :param options: options for the transformation.
+        :type options: Optional[Dict[str, Any]]
+
+        :raises TransformationError: if the supplied node is not an \
+            intrinsic.
+        :raises TransformationError: if the supplied node is not a sum, \
+            minval, or maxval intrinsic.
+        :raises TransformationError: if a valid value for the \
+            dimension argument can't be determined.
+        :raises TransformationError: if the array argument is not an array.
+        :raises TransformationError: if the shape of the array is not \
+            supported.
+        :raises TransformationError: if the array datatype is not \
+            supported.
+
+        '''
+        if not isinstance(node, IntrinsicCall):
+            raise TransformationError(
+                f"Error in {self.name} transformation. The supplied node "
+                f"argument is not an intrinsic, found "
+                f"'{type(node).__name__}'.")
+
+        if node.routine.name.upper() != self._INTRINSIC_NAME:
+            raise TransformationError(
+                f"Error in {self.name} transformation. The supplied node "
+                f"argument is not a {self._INTRINSIC_NAME.lower()} "
+                f"intrinsic, found '{node.routine.name}'.")
+
+        array_ref, dim_ref, _ = self._get_args(node)
+        if dim_ref and not isinstance(dim_ref, (Literal, Reference)):
```

Comment:
Yes, we only allow situations where we can determine the value of dimension at PSyclone processing time. This means we can create a single bespoke loop. If we deal with the case where we don't know the value until run-time the simplest (performant) solution would be to create multiple loops i.e.:

```
dim = expr
if dim ==1 then
  bespoke loop nest for dim==1
elif dim == 2 then
  bespoke loop nest for dim == 2
elif ...
...
end if
```
Clearly this could be implemented but I thought I would just re-use the existing `sum` implementation for the moment. I would suggest adding this is an additional issue (if it is needed for nemo performance).